### PR TITLE
Opentracing library

### PR DESCRIPTION
**Pull request by cdecker**
This is as experimental as it gets, but I had some fun trying to shoehorn a tracing library into CLN. Putting it here so I don't lose it, and hopefully get some feedback on the API.

Ideally I'd love to integrate closer with `io_loop`, namely having it suspend and resume based on the `io_plan` automatically so users don't have to when suspending on a higher level. Then it boils down to starting and ending spans at the entry and exit point of logical calls.

### CODE REVIEW

**Comment by jb55** [comment]
we use ebpf tracepoints in bitcoin, maybe something to check out: 

https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md

super lightweight and has linux kernel support.

**Comment by cdecker** [comment]
Yeah, I was looking into that, however it doesn't mesh too well with the async IO loop we're using, when spans live across several context switches. Ideally I'd like to hook into `io_loop` to suspend and resume spans automatically so instrumenting something on the user side just boils down to using `span_start` and `span_end`.

Either way, the majority of the code here is to manage parent spans anyway and do span lookups by context object, which we'd likely require for eBPF variants too (though we can externalize the reporting of the spans, which is indeed a big win).

**Comment by rustyrussell** [comment]
Bumped to next milestone due to lack of time!

**Comment by cdecker** [comment]
Rebased and fixed up the no-USDT header case, which was causing some issues.

File: common/trace.c [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+#include <stdio.h>
+#include <ccan/short_types/short_types.h>
+#include <ccan/time/time.h>
+#include <ccan/tal/tal.h>
+#include <ccan/tal/str/str.h>
+#include <ccan/str/hex/hex.h>
+#include <common/memleak.h>
+#include <assert.h>
+#include <sodium/randombytes.h>
+#include <common/json_stream.h>
```
Comment (vincenzopalazzo) [inline common/trace.c:13]: needs some more love

```
Cloning into '/home/runner/work/lightning/lightning/external/lowdown'...
cargo build --profile=release --quiet --bin cln-grpc
xgettext plugins/bkpr/statements_gettextgen.po
Submodule path 'external/gheap': checked out '67fc83bc953324f4759e52951921d730d7e65099'
Submodule path 'external/jsmn': checked out '18e9fe42cbfe21d65076f5c77ae2be379ad1270f'
Submodule path 'external/libbacktrace': checked out '2446c66076480ce07a6bd868badcbceb3eeecc2e'
Submodule path 'external/libsodium': checked out '675149b9b8b66ff44152553fb3ebf9858128363d'
Submodule path 'external/libwally-core': checked out 'bb4cd3ac802c7beb58f63307c5ed6ca116cf0dd0'
Submodule 'src/secp256k1' (https://github.com/ElementsProject/secp256k1-zkp.git) registered for path 'external/libwally-core/src/secp256k1'
Cloning into '/home/runner/work/lightning/lightning/external/libwally-core/src/secp256k1'...
common/trace.c:1: includes out of order
#include "config.h"
#include <common/trace.h>
#include <ccan/htable/htable.h>
#include <stdio.h>
#include <ccan/short_types/short_types.h>
#include <ccan/time/time.h>
#include <ccan/tal/tal.h>
#include <ccan/tal/str/str.h>
#include <ccan/str/hex/hex.h>
#include <common/memleak.h>
#include <assert.h>
#include <sodium/randombytes.h>
#include <common/json_stream.h>
#include <sys/sdt.h>
VERSUS
#include "config.h"
#include <assert.h>
#include <ccan/htable/htable.h>
#include <ccan/short_types/short_types.h>
#include <ccan/str/hex/hex.h>
#include <ccan/tal/str/str.h>
#include <ccan/tal/tal.h>
#include <ccan/time/time.h>
#include <common/json_stream.h>
#include <common/memleak.h>
#include <common/trace.h>
#include <sodium/randombytes.h>
#include <stdio.h>
#include <sys/sdt.h>
make: *** [Makefile:482: check-src-include-order/common/trace.c] Error 1
make: *** Waiting for unfinished jobs....
Submodule path 'external/libwally-core/src/secp256k1': checked out 'ff33018fe765df82f8515c564d3fe44d388d3903'
Submodule path 'external/lnprototest': checked out 'dea47c29b5541dbfe7fe53cc2598330e897fa4f4'
Submodule path 'external/lowdown': checked out 'edca6ce6d5336efb147321a43c47a698de41bb7c'
```

**Review by vincenzopalazzo — DISMISSED** [review]

File: common/trace.c [hunk 2]
```diff
@@ -38,7 +39,7 @@ static struct span *active_spans;
 static struct span *current;
 static size_t last_span_id;
 
-static void trace_span_id_serialize(struct span *s, char *dest)
+static void trace_span_id_serialize(struct span *s, char *dest)g
```
Comment (vincenzopalazzo) [inline common/trace.c:42]: there is a `g` at the end of the `static void trace_span_id_serialize(struct span *s, char *dest)`

Maybe we should enable the tracing inside the CI? I could help with some timeout debugging?

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
To enable the tracer we need to enable it at configuration time ? this is really cool!

There is a small typo in the last commit. Then I thought that enabling it in the CI could help in debugging? or this would slow down all the CI?

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/5492/commits/d513df6bb74d93e7de8ce456793708a0300d719e

Sorry for the delay!

P.S: CI needs just a flaky rerun due to a timeout

**Comment by rustyrussell** [comment]
@cdecker Can we have a Changelog-Added and ideally some docs on this?  Would be a great thing to promote for the coming release!

**Comment by cdecker** [comment]
Added a new commit, with both the Changelog and the `tracing.md` document in the docs. The exporter is still a bit minimal, and I'm still trying to build one that can just be connected to an opentracing ingester, but for now `bpftrace` does the trick.

File: doc/developers-guide/tracing.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+CLN will build with tracing support if the necessary headers
+(`sys/sdt.h`) are present during the compilation. For debian and
+ubuntu based systems that is easily achieved by installing
+`systemtap-sdt-dev`:
+
+```bash
+# apt-get install -y systemtap-sdt-dev
+```
+-
```
Comment (vincenzopalazzo) [inline doc/developers-guide/tracing.md:32]:
```suggestion
```

This is odd in the rendered page

**Review by vincenzopalazzo — APPROVED** [review]
LGTM also the last commit, I just have a small comment and then for me we are ready to go 😄

**Review by vincenzopalazzo — APPROVED** [review]
I do not know if you want to squash the last commit, but it is fine for me also in this way.

ACK https://github.com/ElementsProject/lightning/pull/5492/commits/8129d379cc833a40895711bc200de8ecf1b895a0

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/5492/commits/e255c6e394cedaa6e35e8e36695ba43deed449bb

**Review by vincenzopalazzo — COMMENTED** [review]
MH CI is complaining about missing import inside some C tests

```
ld wallet/test/run-db
ld wallet/test/run-psbt_fixup
/usr/bin/ld: wallet/test/run-db.o: in function `wallet_new':
/home/runner/work/lightning/lightning/./wallet/wallet.c:114: undefined reference to `trace_span_start'
/usr/bin/ld: /home/runner/work/lightning/lightning/./wallet/wallet.c:116: undefined reference to `trace_span_end'
/usr/bin/ld: /home/runner/work/lightning/lightning/./wallet/wallet.c:120: undefined reference to `trace_span_start'
/usr/bin/ld: /home/runner/work/lightning/lightning/./wallet/wallet.c:122: undefined reference to `trace_span_end'
/usr/bin/ld: /home/runner/work/lightning/lightning/./wallet/wallet.c:124: undefined reference to `trace_span_start'
/usr/bin/ld: /home/runner/work/lightning/lightning/./wallet/wallet.c:127: undefined reference to `trace_span_end'
collect2: error: ld returned 1 exit status
make: *** [Makefile:665: wallet/test/run-db] Error 1
make: *** Waiting for unfinished jobs....
rm external/build-x86_64-linux-gnu/libwally-core-build/src/secp256k1/libsecp256k1.la
```

**Comment by rustyrussell** [comment]
Since I had to fix the test build breakage anyway, I reshuffled the commits to make it bisectable (moving the configure test to the first commit).  I checked that there's no net change, so:

Ack 5b27344a9301c1ad05751fc3856cedda63cc02d3

**Pull request merged**